_posts/2016-05-22-th1eadin0g1.markdown

```diff
@@ -0,0 +1,202 @@
+
+* **Using `NSInvocationOperation`** - This is actually what I recommended in my blog post in 2012, it's essentially a simple wrapper around `NSTask` which calls a specific method.
+* **Using `NSBlockOperation`** - This is a nice option when you want to run some code from a block. More on blocks later. 
+
+`NSOperation`s on their own are  pretty simple objects, but they allow you to define dependency graphs. So you could have one operation taking the JPG from the file system, then another operation that takes the image data and converts it in an `NSImage`, then the latter can depend on the former. Now you have some heirarchy. That's nice.
+
+No discussion about `NSOperation` would be complete without `NSOperationQueue`. This is the abstraction you use to limit how many operations can be running at the same time. Say you had a lot of images to download, you could use an operation queue with a limit of three concurrent operations. They also provide an easy way to determine priority too, so you could have a higher priority queue for images you would see, then a lower priority queue for images that are not on the screen at the moment.
+
+This is only a small overview, for more information, I strongly recommend reading [NSHipster's NSOperation](http://nshipster.com/nsoperation/) post. 
+
+### Grand Central Dispatch (GCD)
+
+This is the most used tool in modern Objective-C threading. GCD relies on blocks. As I know some android people are reading this, a block is a way of treating code in similar way that you would variables. These can be passed between methods, a very common case is callbacks. They let you do some work, then let something know that you've completed the work. The advantage being that you can keep all the code in one place.
+
+Back to GDC. The rough gist, is that the OS should handle creating threads, and you feed it bits of code to run. We call this `dispatch`ing the block. 
+
+Going back to our real example, in 2012, I recommended `NSOperation`s, in 2014, I migrated that code to GCD. [It now looks like this](https://github.com/artsy/energy/blob/9fb6faef18361772bde47513f44e842c592641fc/Classes/Views/Grid%20View%20%26%20Items/ARGridView.m#L309-L342):
+
+```objc
+- (void)setImageAsyncAtPath:(NSString *)imageAddress backupURL:(NSURL *)backupURL forGridCell:(ARGridViewCell *)cell asButton:(BOOL)asButton
+{
+    ar_dispatch_async(^{
+        // don't load if it's on a different cell
+        if (![cell.imagePath isEqual:imageAddress]) return ;
+
+        UIImage *thumbnail = [... load data from the disk, and turn into an image]
+
+        ar_dispatch_main_queue(^{
+
+            // Double check that during the decoding the cell's not been re-used
+            if ([cell.imagePath isEqual:imageAddress] && thumbnail) {
+                [cell setImage:thumbnail];
+                
+                [...]
+            }
+        });
+    });
+}
```

Comment:
I would replace the `ar_` specific GCD helpers by the real ones, that will make it easier to lookup docs.